packages/zoe/zoe.js

```diff
@@ -42,137 +35,241 @@ const makeZoe = async (additionalEndowments = {}) => {
+  const {
+    installationTable,
+    instanceTable,
+    offerTable,
+    assayTable,
+  } = makeTables();
 
-  // Zoe has two different facets: the public Zoe service and the
-  // contract facet. Neither facet should give direct access
-  // to the `adminState`.
+  // Helper functions
+  const getAssaysFromPayoutRules = payoutRules =>
+    payoutRules.map(payoutRule => payoutRule.units.label.assay);
+
+  // In the future, this function will take `offerHandles` and
+  // `assays` as parameters.
+  const completeOffers = offerHandles => {
+    const { inactive } = offerTable.getOfferStatuses(offerHandles);
+    if (inactive.length > 0) {
+      throw new Error(`offer has already completed`);
+    }
+
+    // In the future, when `assays` is a parameter, these next two
+    // lines can be deleted
+    const payoutRules = offerTable.getPayoutRules(offerHandles[0]);
+    const assays = getAssaysFromPayoutRules(payoutRules);
+
+    const unitMatrix = offerTable.getUnitMatrix(offerHandles, assays);
+    const payoutPromises = offerTable.getPayoutPromises(offerHandles);
+
+    // Remove the offers from the offerTable so that they are no
+    // longer active.
+    offerTable.deleteOffers(offerHandles);
+
+    // Resolve the payout promises with the payouts
+    const pursesP = assayTable.getPursesForAssays(assays);
+    Promise.all(pursesP).then(purses => {
+      for (let i = 0; i < offerHandles.length; i += 1) {
+        const unitsForOffer = unitMatrix[i];
+        // This Promise.all will be taken out in a later PR.
+        const payout = Promise.all(
+          unitsForOffer.map((units, j) =>
+            E(purses[j]).withdraw(units, 'payout'),
+          ),
+        );
+        payoutPromises[i].res(payout);
+      }
+    });
+  };
 
-  // The contract facet is what is accessible to the smart contract
-  // instance and is remade for each instance. The contract at no time
-  // has access to the users' payments or the Zoe purses, or any of
-  // the `adminState` of Zoe. The contract can only do a few things
-  // through the Zoe contract fact. It can propose a reallocation of
-  // extents, complete an offer, and interestingly, can create a new
-  // offer itself for record-keeping and other various purposes.
+  // Create payoutRules in which nothing is offered and anything
+  // greater than or equal to 'empty' satisfies offer safety.
+  const makeEmptyPayoutRules = unitOpsArray =>
+    unitOpsArray.map(unitOps =>
+      harden({
+        kind: 'wantAtLeast',
+        units: unitOps.empty(),
+      }),
+    );
+
+  // Make a Zoe invite with an extent that is a mix of credible
+  // information from Zoe (the `handle` and `instanceHandle`) and
+  // other information defined by the smart contract. Note that the
+  // smart contract cannot override or change the values of `handle`
+  // and `instanceHandle`.
+  const makeInvite = (
+    instanceHandle,
+    seat,
+    contractDefinedExtent = harden({}),
```

Comment:
Hmm, if we didn't flatten it, what would we call the property that it is under? Flattening seemed like the natural thing to do, to me. It seemed easier to look at. 

By the way, if we choose to change this, it will break the current invites as used by the covered call and we should do it as a separate PR.

Reply:
sure, we can change it. I changed the name to `customProperties`